src/compiler/program.ts

```diff
@@ -748,7 +748,21 @@ namespace ts {
         }
 
         if (!tryReuseStructureFromOldProgram()) {
-            forEach(rootNames, name => processRootFile(name, /*isDefaultLib*/ false));
+            let programSize = 0;
+            for (const name of rootNames) {
+                const path = toPath(name, currentDirectory, getCanonicalFileName);
+                if (programSize <= maxProgramSize) {

```

Comment:
We [discussed a compiler option](https://github.com/Microsoft/TypeScript/issues/7444#issuecomment-194491024) to disable the size check for cases where the project is legitimately large. I don't see that option in this code review.